### PR TITLE
Add unit tests for completion models

### PR DESCRIPTION
**Pull request by rcorre**
This establishes a pattern that can probably be used to test any of the
completion models.

See #999.

Let me know if this pattern looks good, and I'll start testing the other models.

### CODE REVIEW

**Comment by rcorre** [comment]
Just realized there is an easy(ish) way to stub `cmdutils.cmd_dict`. Gonna do that, ignore this for now.

File: tests/unit/completion/test_models.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    monkeypatch.setattr('qutebrowser.completion.models.miscmodels.cmdutils',
+                        cmd_utils)
+    config_stub.data['aliases'] = {'rock': 'roll'}
+    key_config_stub.set_bindings_for('normal', {'s': 'stop', 'rr': 'roll'})
+    actual = _get_completions(miscmodels.CommandCompletionModel())
+    assert "Commands" in actual
+    assert actual["Commands"] == [('drop', 'drop all user data', ''),
+                                  ('rock', "Alias for 'roll'", ''),
+                                  ('roll', 'never gonna give you up', 'rr'),
+                                  ('stop', 'stop qutebrowser', 's')]
```
Comment (The-Compiler) [inline tests/unit/completion/test_models.py]: :laughing: :+1:

File: tests/unit/completion/test_models.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    """Collect all the completion entries of a model, organized by category.
+
+    The result is a dict of form:
+    {
+        CategoryName: [(name, desc, misc), ...],
+        CategoryName: [(name, desc, misc), ...],
+        ...
+    }
+    """
+    completions = {}
```
Comment (The-Compiler) [inline tests/unit/completion/test_models.py]: Maybe this should be a [`collections.OrderedDict`](https://docs.python.org/3.5/library/collections.html#collections.OrderedDict) so future tests could also check the order of completions?

File: tests/unit/completion/test_models.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        'roll': stubs.FakeCommand(name='roll', desc='never gonna give you up'),
+        'hide': stubs.FakeCommand(name='hide', hide=True),
+        'depr': stubs.FakeCommand(name='depr', deprecated=True),
+    })
+    monkeypatch.setattr('qutebrowser.completion.models.miscmodels.cmdutils',
+                        cmd_utils)
+    config_stub.data['aliases'] = {'rock': 'roll'}
+    key_config_stub.set_bindings_for('normal', {'s': 'stop', 'rr': 'roll'})
+    actual = _get_completions(miscmodels.CommandCompletionModel())
+    assert "Commands" in actual
```
Comment (The-Compiler) [inline tests/unit/completion/test_models.py]: `assert actual.keys() == ["Commands"]` maybe to ensure there are no other categories?

File: tests/unit/completion/test_models.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with qutebrowser.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Tests for completion models."""
+
+import pytest
```
Comment (The-Compiler) [inline tests/unit/completion/test_models.py]: unused import

File: tests/unit/completion/test_models.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+from qutebrowser.completion.models import miscmodels
+from qutebrowser.commands import cmdutils
```
Comment (The-Compiler) [inline tests/unit/completion/test_models.py]: unused import

**Comment by The-Compiler** [comment]
Nice! I added a few comments, but looks like a simple good way to test completion models indeed.

File: tests/unit/completion/test_models.py [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+def _get_completions(model):
+    """Collect all the completion entries of a model, organized by category.
+
+    The result is a list of form:
+    [
+        (CategoryName: [(name, desc, misc), ...]),
+        (CategoryName: [(name, desc, misc), ...]),
+        ...
+    ]
```
Comment (rcorre) [inline tests/unit/completion/test_models.py]: Just went with a list for the whole thing here so you can check the order of the categories too.

**Comment by The-Compiler** [comment]
Is this ready to review/merge?

**Comment by rcorre** [comment]
At this point I might as well finish the rest of the models. I think I just have the config models left. I've been grouping the models into separate commits in case its easier to review that way.

**Comment by rcorre** [comment]
@The-Compiler assuming the CI tests pass, I think this is ready to go.

**Comment by rcorre** [comment]
Looks like something is awry with the [osx env](https://travis-ci.org/The-Compiler/qutebrowser/jobs/140439234)

**Comment by The-Compiler** [comment]
Sorry for the delay - after failing to properly fix this, I now marked OS X tests as allowed to fail again for the time being... Reviewing this now!

File: tests/helpers/stubs.py [hunk 7]
```diff
@@ -335,6 +349,44 @@ def isActive(self):
...
+    def __init__(self, *entries):
+        self.values = collections.OrderedDict()
+        self.descriptions = {}
+        for name, value, desc in entries:
+            self.values[name] = value
+            self.descriptions[name] = desc
+
+    def __iter__(self):
+        """Iterate over all set values."""
+        return self.values.__iter__()
```
Comment (The-Compiler) [inline tests/helpers/stubs.py]: nitpick: This should be `iter(self.values)` (just like you wouldn't do `self.values.__getitem__(key)` below) :wink:

File: tests/helpers/stubs.py [hunk 8]
```diff
@@ -335,6 +349,44 @@ def isActive(self):
...
+
+    def __getitem__(self, key):
+        return self.values[key]
+
+
+class FakeSettingValue:
```
Comment (The-Compiler) [inline tests/helpers/stubs.py]: Not sure if it really pays off to have a fake for this - couldn't you easily use the real `SettingValue` object without pulling it too much unrelated code?

File: tests/helpers/stubs.py [hunk 9]
```diff
@@ -400,9 +452,114 @@ def set_bindings_for(self, section, bindings):
         self.bindings[section] = bindings
 
 
+class FakeHistoryEntry:
```
Comment (The-Compiler) [inline tests/helpers/stubs.py]: Same as above, I think you could just use `webkit.history.Entry` here?

File: tests/unit/completion/test_models.py [hunk 10]
```diff
@@ -0,0 +1,386 @@
...
+
+def _mock_view_index(model, category_idx, child_idx):
+    """Create a tree view from a model and set the current index.
+
+    Args:
+        model: model to create a fake view for.
+        category_idx: index of the category to select.
+        child_idx: index of the child item under that category to select.
+    """
+    view = QTreeView()
```
Comment (The-Compiler) [inline tests/unit/completion/test_models.py]: Please get the `qtbot` fixture and do `qtbot.add_widget(view)` so the view correctly gets cleaned up after the test is over.

File: tests/unit/completion/test_models.py [hunk 11]
```diff
@@ -0,0 +1,386 @@
...
+            ('11', 'Default value', ''),
+        ]),
+        ("Completions", [
+            ('0', '', ''),
+            ('11', '', ''),
+        ])
+    ]
+
+
+def _get_completions(model):
```
Comment (The-Compiler) [inline tests/unit/completion/test_models.py]: I think all those utility functions should be at the top of the file rather than the bottom. Makes it easier to read the file from top to bottom :wink:

**Comment by The-Compiler** [comment]
Nice work overall!

As mentioned in the line comments I'm somewhat wary about all the mocking though. It's always a trade-off, as it increases test maintainance costs for future changes and also opens the possibility of tests passing but the real code failing because of a discrepancy there.

While it makes sense to mock bigger things off which you can't easily instantiate (say, a `WebView` because of all its dependencies), for smaller things like the ones you creating mocks for, I think it's perfectly fine to use the real thing if possible.

Could you take another look at what actually makes sense to mock out, and where it'd be simpler to use the real thing?

File: tests/helpers/stubs.py (hunk 8 above)
Comment (rcorre) [inline tests/helpers/stubs.py]: > Though I guess this makes it easier to customize valid_values and default for tests?

Exactly. If I use a real `SettingValue`, then I have to use a real `ConfigType`, and my test becomes flaky with regards to changes in configtypes (e.g. if someone added an extra option to `IgnoreCase`).

**Comment by rcorre** [comment]
> Could you take another look at what actually makes sense to mock out, and where it'd be simpler to use the real thing?

Definitely. It depends on whether you really think of these 'unit' tests as 'isolation' tests, or whether they could have some elements of an integration test. I was leaning towards the former, but it definitely feels like I mocked out too much.

In particular, it might make sense to use a real `Command` object. One refactoring I wanted to do was remove `Command.completion` and just rely directly on the `completion` of the `ArgInfo`. Using a `FakeCommand`, my tests wouldn't catch that. Actually I can't even remember now why I didn't use a real `Command`... it seemed like there was a good reason because I definitely thought of this at the time.

**Comment by rcorre** [comment]
Rebased on master to pick up the osx test 'fix'. The last 4 commits are from this code review. I've still got some more un-mocking to do, I think.

**Comment by The-Compiler** [comment]
> Definitely. It depends on whether you really think of these 'unit' tests as 'isolation' tests, or whether they could have some elements of an integration test. I was leaning towards the former, but it definitely feels like I mocked out too much.

Yeah, I'm usually leaning a bit more towards the latter.

I think the benefit is clear: More straightforward and often more maintainable tests.

The drawbacks I can think of:

- Slightly slower tests - I don't think it makes much of a difference, unless
  you have things like network access, which should definitely be mocked out.
- Unrelated code changes cause unit tests to fail - I think here the benefits
  simply outweight this drawback, and if you mocked things out, you need to
  take care of keeping the mocks up-to-date as well
- Coverage can be a bit skewed, i.e. a module can appear covered even though
  it's just from other tests "smearing" to that module, and it's not actually
  tested well. To take care of that, I added a mode to run test files
  one-by-one to `check_coverage.py`, though I haven't actually checked if that
  still works for a while :laughing:

File: tests/helpers/stubs.py (hunk 7 above)
Comment (The-Compiler) [inline tests/helpers/stubs.py]: Oh, my bad! :laughing: Thanks for fixing it!

File: tests/unit/completion/test_models.py [hunk 12]
```diff
@@ -0,0 +1,388 @@
...
+
+
+def test_url_completion_delete_bookmark(config_stub, web_history, quickmarks,
+                                        bookmarks, qtbot):
+    """Test deleting a bookmark from the url completion model."""
+    config_stub.data['completion'] = {'timestamp-format': '%Y-%m-%d',
+                                      'web-history-max-items': 2}
+    model = urlmodel.UrlCompletionModel()
+    # delete item (1, 0) -> (bookmarks, 'https://github.com' )
+    view = _mock_view_index(model, 1, 0)
```
Comment (The-Compiler) [inline tests/unit/completion/test_models.py]: I think you need to do `qtbot.add_widget(view)` here too.

File: tests/unit/completion/test_models.py [hunk 13]
```diff
@@ -0,0 +1,388 @@
...
+
+
+def test_url_completion_delete_quickmark(config_stub, web_history, quickmarks,
+                                        bookmarks, qtbot):
+    """Test deleting a bookmark from the url completion model."""
+    config_stub.data['completion'] = {'timestamp-format': '%Y-%m-%d',
+                                      'web-history-max-items': 2}
+    model = urlmodel.UrlCompletionModel()
+    # delete item (0, 1) -> (quickmarks, 'ddg' )
+    view = _mock_view_index(model, 0, 1)
```
Comment (The-Compiler) [inline tests/unit/completion/test_models.py]: I think you need to do `qtbot.add_widget(view)` here too.

**Comment by rcorre** [comment]
I'm pretty much in complete agreement about mocking. I've seen tests that mock out so much they don't even seem to be testing anything ...

I managed to eliminate the fake setting sections and values; I kept around `FakeConfigType` because its an easy way to mock completions.

I remember now why I didn't use a real `Command`: the constructor takes the handler method and tears it apart to populate its fields, and that's way more than I wanted to test here. I can take a look at `test_cmdutils`/`test_argparser` and see how rigorous they are.

I've got a test for the `Completer` in the pipeline as well -- should I push that up here or would you rather get this through and review that in another PR?

**Comment by The-Compiler** [comment]
Let's get this in first - thanks!
